### PR TITLE
Issue #6563: Fix Invalid Doc

### LINKED ISSUE #6563: Site files are not schema-valid
Currently the `validate` property of the `maven-site-plugin` is disabled because of a bug in `xml-maven-plugin` that was fixed in `1.0.2`.
Since checkstyle in the meantime uses that version, the `validate` could be enabled again.
Unfortunately in the long time it was off, some invalid markup made it into the site files.

I'd like to suggest to reenable the validation and fix the invalid parts.
In detail this would be:
- text directly within `<section>` needs to be wrapped in paragraphs
- block tags like `<div>` or `<ul>` must not occur within paragraphs, but besides
- `<source>` tag does not allow `class` attribute but instead you have to wrap a `<div>` around it and adapt the CSS accordingly to have the wanted effect
- Within `<body>` only `<section>` is allowed, not `<div>` like in the release notes, but via CSS placement it can still look the same in the end even with the release date within the section which is semantically cleaner anyway as it belongs to the respective version (this also needs to be changed in the release notes generator template)
- Not exactly due to schema-validity, but because the `wrap-content` style needs to be changed anyway I also suggest to change it a tad bit, so that not `white-space: pre-wrap` but `white-space: pre-line` is used. 
With the former all whitespace is preserved, but if space gets narrow (too long text, too much zoom) the text is automatically wrapped.
With the latter line-breaks are preserved, but leading spaces are stripped from each line.
This means, that you can then properly indent the code in the sources or you can have them at column 0 or anywhere in between and it will look the same on the final page so it doesn't matter if someone indents it accidentally for example unlike currently.
This also means if you want to have visual indentation on the page, you have to use a non-breaking space as `&#xa0;`.
Another big plus you get which is why I looked into it actually is, that also trailing spaces get removed which looks nicer in the result.
grafik
vs.
grafik
And in the code
grafik
vs.
grafik

### PR DESCRIPTION
**Pull request by Vampire**
- re-enable maven-site-plugin validation as underlying problem was fixed in xml-maven-plugin 1.0.2
- fix invalid markup

Fixes #6563

### CODE REVIEW

File: src/xdocs/beginning_development.xml [hunk 1]
```diff
@@ -31,15 +31,21 @@
         <a href="https://help.github.com/articles/fork-a-repo/"> here</a><br/>
         3. Checkout the current source code from: https://github.com/you_user_name/checkstyle/<br/>
         by running
-        <source class="wrap-content">
-git clone git@github.com:you_user_name/checkstyle.git
+      </p>
+      <div class="wrap-content">
+        <source>
+            git clone git@github.com:you_user_name/checkstyle.git
```
Comment (rnveach) [inline src/xdocs/beginning_development.xml]: no indentation is intended.
This is the same for all similar to this. This is to help prevent hitting the 100 character line limit.

File: src/xdocs/releasenotes.xml [hunk 2]
```diff
@@ -16,8 +16,8 @@
   <body>
     <!-- placeholder for a new section -->
 
-    <div class="releaseDate">25.02.2019</div>
     <section name="Release 8.18">
+      <div class="releaseDate">25.02.2019</div>
```
Comment (rnveach) [inline src/xdocs/releasenotes.xml]: This is a generated section and cannot be modified easily.
Please explain why this change is needed, and if it is valid, it will also have to be updated at https://github.com/checkstyle/contribution/blob/master/releasenotes-builder/src/main/resources/com/github/checkstyle/templates/xdoc_freemarker.template#L2 .

File: src/xdocs/beginning_development.xml [hunk 3]
```diff
@@ -31,15 +31,21 @@
         <a href="https://help.github.com/articles/fork-a-repo/"> here</a><br/>
         3. Checkout the current source code from: https://github.com/you_user_name/checkstyle/<br/>
         by running
-        <source class="wrap-content">
-git clone git@github.com:you_user_name/checkstyle.git
+      </p>
+      <div class="wrap-content">
+        <source>
```
Comment (rnveach) [inline src/xdocs/beginning_development.xml]: Why the change from `<source>` to `<div><source>`?

File: .ci/jsoref-spellchecker/whitelist.words [hunk 4]
```diff
@@ -1350,6 +1350,7 @@ writingjavadocchecks
 writinglisteners
 Wsdl
 www
+xa
```
Comment (rnveach) [inline .ci/jsoref-spellchecker/whitelist.words]: This looks like it is from ` &#xa0;` which is the same as `&nbsp;`. We should use `&nbsp;` since it is already whitelisted.

**Comment by rnveach** [comment]
I also question if this is really a `doc` type change. @romani ping

File: src/xdocs/beginning_development.xml (hunk 1 above)
Comment (Vampire) [reply src/xdocs/beginning_development.xml]: Well, with my changes you can indent it or not, it doesn't make any difference in the final result.
I can of course unindent the parts again, but don't you think it looks cleaner indented?

File: src/xdocs/releasenotes.xml (hunk 2 above)
Comment (Vampire) [reply src/xdocs/releasenotes.xml]: The change is necessary, because the only valid content of `<body>` tag is `<section>`, so the `<div>` cannot be before the section which is why I put it inside and moved it to the proper place via CSS.
I was not aware that this file is generated, I now also adapted the freemarker template and the tests, the generation produces the proper result here now, I'll create a PR to the `contribution` repo when we agreed here.

File: src/xdocs/beginning_development.xml (hunk 3 above)
Comment (Vampire) [reply src/xdocs/beginning_development.xml]: Because `<source>` must not have a `class` attribute.

Comment (rnveach) [reply src/xdocs/beginning_development.xml]: Why? Is there some place that states it is not valid? We use it already and have had no issues.

File: src/xdocs/releasenotes.xml (hunk 2 above)
Comment (rnveach) [reply src/xdocs/releasenotes.xml]: @romani ping

File: src/xdocs/beginning_development.xml (hunk 1 above)
Comment (rnveach) [reply src/xdocs/beginning_development.xml]: It may look neater in the source code, but it also carried through to the web page. And we have add issues with line lengths as stated before.
I think we would like to keep them unindented.

Comment (Vampire) [reply src/xdocs/beginning_development.xml]: I don't get what you mean.
The `MissingOverride` example is not marked with `class` `wrap-content` which is why it is strangely indented.
If it were marked with `class` `wrap-content` and my changes applied, the leading whitespace is stripped and so it would look proper, no matter whether you indent it in the source or not.

Again, with my changes the `.wrap-content` class changed from `white-space: pre-wrap` to `white-space: pre-line` where the leading whitespace is stripped in the display, so it makes absolutely no difference in the end-result whether you indent the example in the source or not.

File: src/xdocs/releasenotes.xml (hunk 2 above)
Comment (Vampire) [reply src/xdocs/releasenotes.xml]: Just to mention it again, if you are not ok with this, you can never enable the validation option of the maven-site-plugin, as the `<div>` as child of the `<body>` is invalid and thus will not validate.
Up to now the validation option was disabled due to a bug linked in the `pom.xml` or rather an underlying bug in the `maven-xml-plugin` that since was fixed.

File: src/xdocs/beginning_development.xml (hunk 3 above)
Comment (Vampire) [reply src/xdocs/beginning_development.xml]: Because the validation of the maven-site-plugin is disabled because of a bug that is fixed now.
If you switch on validation and do not do the changes I made, the pages will not validate.
I basically mostly did changes that were necessary to make the pages validate properly.
It worked until now because validation was turned of because of said bug and browsers are very lenient in parsing HTML and correcting errors in it to have a better experience for users that browse to pages that are acutally not valid.

File: src/xdocs/beginning_development.xml (hunk 1 above)
Comment (rnveach) [reply src/xdocs/beginning_development.xml]: > wrap-content which is why it is strangely indented.

@Vampire I didn't know thats what the `wrap-content` changes were for.
How will it handle consecutive lines? If it removes all indentation then that will cause its own issue.

Instead of:
````
<module name="MissingOverride">
  <property name="javaFiveCompatibility"
    value="true"/>
</module>
````

We will  get:
````
<module name="MissingOverride">
<property name="javaFiveCompatibility"
value="true"/>
</module>
````

Which is not easy to read.
It may also make the display weird if we have a long line that we need to break apart more as it gets deeply nested. It looks like `wrap-content` still keeps the line breaks, so deeply nested sources will have very short lines in source and display.
Unless you can provide another alternative, then I think what we had before works better for us.

File: src/xdocs/releasenotes.xml (hunk 2 above)
Comment (rnveach) [reply src/xdocs/releasenotes.xml]: Can you provide screenshot of how page will look with your changes? If it is no difference, then I am fine.
I would like @romani to confirm before any changes are made to contribution.

File: src/xdocs/beginning_development.xml (hunk 3 above)
Comment (rnveach) [reply src/xdocs/beginning_development.xml]: Ok thanks.
If there is no or minor change in display because of these 2 line changes then I am fine. Can you provide a screenshot?

Comment (romani) [reply src/xdocs/beginning_development.xml]: @Vampire , thanks a lot for your desire to help, I really appreciate this, but lets follow rules of our project, we will guide you. Sorry if extra work might be required.

first of all this is not a "doc:" minor change, please create issue on it and lets describe all reasons/nuances in issue, not a PR comments.

File: src/xdocs/beginning_development.xml (hunk 1 above)
Comment (romani) [reply src/xdocs/beginning_development.xml]: we did no proper indentation to keep our source code render in HTML with no indentation.
I do not like our approach also, but I do not know for now how to make it better. If you know please share an approach and lets move this to separate commit/PR, or it is caused by plugin validation?.

File: src/xdocs/releasenotes.xml (hunk 2 above)
Comment (romani) [reply src/xdocs/releasenotes.xml]: @Vampire , please share a screenshot that it renders fine, command to generate site is `mvn site -Pno-validations`.

File: .ci/jsoref-spellchecker/whitelist.words (hunk 4 above)
Comment (romani) [reply .ci/jsoref-spellchecker/whitelist.words]: @Vampire ,
I am not sure what for you try to use https://en.wikipedia.org/wiki/Non-breaking_space after `\` in command lines.
On small screens commands are look not good, but if we want to make them better  it is better to use such symbol in even before `\`.
Please move this update to separate issue/PR. I doubt it is related to plugin validation, correct me if I am wrong.

**Comment by romani** [comment]
@Vampire , @rnveach 
I replied to conversations.

File: src/xdocs/beginning_development.xml (hunk 1 above)
Comment (Vampire) [reply src/xdocs/beginning_development.xml]: The `.source` style for the `<source>` tags has `white-space: nowrap;` which means no automatic wrapping is done by the browser.
This then caused #3902.
To fix this the `wrap-content` class was introduced which set `white-space: pre-wrap !important;` for selected cases.
This caused the whitespaces to be preserved but if necessary line-wrapping to be done and thus also required all relevant code parts to be unindented.
Additionally as it preserves linebreaks, it looks like this (extra line in the end):
grafik
I wanted to look nicer, like this:
grafik
Due to that part of my changes is to use `white-space: pre-line !important;` instead for the `pre` tag. As it strips the whitespace, the strange extra line is gone without the need to always have the `</source>` tag on the last line of the actual content which would be an alternative. A side-effect of this is, that the code parts can be indented properly in the source, as it causes the whitespaces to be stripped, but newlines to be preserved and still wrap if necessary.

For the intentional indenting that you stripped in your "We will get" example are the non-breaking spaces.
Also, this is only for the parts marked with the `wrap-content` class.

File: src/xdocs/releasenotes.xml (hunk 2 above)
Comment (Vampire) [reply src/xdocs/releasenotes.xml]: It will look exactly like currently, one pixel give or take.

File: src/xdocs/beginning_development.xml (hunk 3 above)
Comment (Vampire) [reply src/xdocs/beginning_development.xml]: It looks the same, screenshot is in the comment above I just made.

Comment (Vampire) [reply src/xdocs/beginning_development.xml]: Ah, ok, didn't know `doc:` is only for minor doc changes.
Well, in effect they actually **are** minor, but I'll open an issue later, gotta go now.

File: .ci/jsoref-spellchecker/whitelist.words (hunk 4 above)
Comment (Vampire) [reply .ci/jsoref-spellchecker/whitelist.words]: They are not for nicer look, but to get indentation with the `pre-line` style which strips leading whitespace.
For getting the markup valid I also had to change the styles a bit and tweaked it alongside as described in my other comment

**Comment by rnveach** [comment]
Generated Site: http://rveach.no-ip.org/checkstyle/regression/reports/fix-invalid-doc/

**Comment by Vampire** [comment]
Or http://kautler.net/cs/

File: src/xdocs/releasenotes.xml (hunk 2 above)
Comment (rnveach) [reply src/xdocs/releasenotes.xml]: http://rveach.no-ip.org/checkstyle/regression/reports/fix-invalid-doc/releasenotes.html

Output looks the same. When PR is ready to be merged, we will have to create PR for other repo.
I will this as unresolved to remind us.

Comment (Vampire) [reply src/xdocs/releasenotes.xml]: I have the respective PR commit for PR ready already, I could open PR to the other repo as reminder if you want.

File: src/xdocs/beginning_development.xml (hunk 3 above)
Comment (rnveach) [reply src/xdocs/beginning_development.xml]: http://rveach.no-ip.org/checkstyle/regression/reports/fix-invalid-doc/beginning_development.html#Before_development

There is a minor change here. Box shrunk compared to original. (I use firefox)
http://checkstyle.sourceforge.net/beginning_development.html#Before_development

I am ok with the shrinkage. It looks like excess space anyways that we didn't need.
I will mark this resolved.

Comment (rnveach) [reply src/xdocs/beginning_development.xml]: @Vampire 
> Because <source> must not have a class attribute.
> Why? Is there some place that states it is not valid?
> Because the validation of the maven-site-plugin is disabled 

Do you know how maven-site-plugin determines what children are acceptable and which aren't? Is it a fixed list built into it's code?

Comment (Vampire) [reply src/xdocs/beginning_development.xml]: > Do you know how maven-site-plugin determines what children are acceptable and which aren't? Is it a fixed list built into it's code?

It is the same children your IDE or any other proper XML editor allows:
grafik

It is defined in the XDoc 2.0 XML schema referenced in the root tag of the file.

File: .ci/jsoref-spellchecker/whitelist.words (hunk 4 above)
Comment (rnveach) [reply .ci/jsoref-spellchecker/whitelist.words]: My original complaint is not valid. I will leave @romani to discuss this indentation usage.
I will mark this resolved.

File: src/xdocs/beginning_development.xml (hunk 1 above)
Comment (rnveach) [reply src/xdocs/beginning_development.xml]: I wasn't aware the changes to the styles removed the extra spaces in front.
I also see now this wasn't done for all sources, just the wrapped content ones.

I did however like the no indentation for these items. It made seeing the internal flow easier as it stood out from the actual xml. It also allowed wraps and non-wraps to have the same indentation make up so you wouldn't accidentally add the indentation to one when it shouldn't have. I also can't say I'm a fan of the `&#xa0;` just because it isn't a common thing I've seen.

I will leave this up to @romani to decide on. Issue still isn't approved.

**Comment by rnveach** [comment]
@Vampire sorry, I didn't see the original link.
I would keep an eye on PRs titled `aligned javadoc/xdoc` being merged as we are re-structuring the javadocs and xdocs which could cause hidden conflicts with your PR if it is not rebased before merge.

**Comment by rnveach** [comment]
@romani I have reviewed the PR and mostly ok with it except for https://github.com/checkstyle/checkstyle/pull/6514#discussion_r264501145 which I will leave you to decide on.
Issue isn't approved so CI is failing just for that.

**Comment by rnveach** [comment]
@romani @Vampire 
The extra PR to contribution talked about at https://github.com/checkstyle/checkstyle/pull/6514#discussion_r262939233 has to be done before the next release.
Since you have made the commit already @Vampire , please submit it.

**Comment by Vampire** [comment]
Sure: checkstyle/contribution#359
